### ISSUE TITLE
A dataset and architecture for visual reasoning with a working memory

### ISSUE THREAD

**Issue by tenaflyyy** [tenaflyyy/Papers#2]
## Abstract
- Challenge
   - Reason about events that occur in complex, changing visual stimuli such as in **video analysis** or game play. 
- Proposed Method
   - The paper developed an vqa dataset termed **COG** , which addresses many of the problems relating to visual and logical reasoning and memory. 
    - Propose **a deep learning architecture** that performs on other diagnostic VQA datasets (i.e. CLEVR) as well as COG. 
    - After training, the network can **zero-shot generalize** to many new tasks. 
## Details
- Introduction
   - Create an artiﬁcial dataset *COG*, whitch is Based on a programmatic language that builds a battery of task triplets: an **image sequence**, a **verbal instruction**, and a **sequence of correct answers**.
   - Introduce a multi-modal **recurrent architecture** for visual reasoning with memory. The network combines **semantic** and **visual** *modules* with a **stateful controller** that modulates visual attention and memory in order to correctly perform a visual task. Tasks in the COG dataset involve objects shown in the current image and in previous images. 
![qq 20180324142358](https://user-images.githubusercontent.com/21164307/37861179-1c0ba3f2-2f6f-11e8-922c-4381715f1f7b.png)
- Contributions 
   - COG dataset
      - The dataset contains **triplets** of a task instruction, sequences of synthetic images, and sequences of target responses. Each image consists of simple objects that vary in color, shape, and location. There are *19 possible colors* and *33 possible shapes* (6 geometric shapes and 26 lower-case English letters). The network needs to **generate a verbal or pointing response** for every image. 
   - Network
      - The network contains four major systems
![qq 20180324143824](https://user-images.githubusercontent.com/21164307/37861257-0b4366a2-2f71-11e8-8c48-f11f18e9e17b.png)
      - The **visual system** processes the images. The inputs are 112*112 images, processed by 4 conv layers with 32,64,64,128 feature maps respectively. Each convolutional layer employs 3×3 kernels and is followed by a 2 × 2 max-pooling layer, batch-normalization , and a rectiﬁed-linear activation function.
      - The **semantic system** processes the task instructions. Each word is mapped to a 64-d embedding vector, then sequentially fed into an 128-unit bidirectional LSTM . The outputs of the bidirectional LSTM for all words form a semantic memory of size (nword,128 ), where nword is the number of words in the instruction, and 128 is the dimension of the output vector. then excute self-attention.
      - The **visual short-term memory system** maintains the processed visual information, and provides outputs that guide the pointing response. This module is an extension of a 2-d convolutional LSTM network  consists of a number of 2-D feature maps, while the input and output connections are both convolutional. There is currently no recurrent connections within the vSTM module besides the forget gate. 
      - The **control system** integrates converging information from all other systems, uses several attention and gating mechanisms to regulate how other systems process inputs and generate outputs, and provides verbal outputs. The controller is a Gated Recurrent Unit (GRU) network. At each pondering step, the post-attention activity of the top visual layer is processed through a 128-unit fully connected layer, concatenated with the retrieved semantic memory and the vSTM module output, then fed into the controller.
      - The network is allowed multiple time steps to “ponder” about each image , giving it the potential to solve multi-step reasoning problems naturally through iteration.
## Experiments
- Perform on CLEVER
- Perform on COG
